### PR TITLE
qtscrcpy: init at 2.2.1

### PR DESCRIPTION
**Pull request by Daru-san**
## Description of changes

Adds qt-scrcpy, a qt frontend for scrcpy.
Homepage : https://github.com/barry-ran/QtScrcpy

Build unsuccessful at the moment, help would be great :smile: 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Daru-san** [comment]
Currently stuck getting this error:
```
┃        > /build/source/QtScrcpy/util/mousetap/xmousetap.cpp: In function 'void find_grab_window_recursive(xcb_connection_t*, xcb_window_t, QRect, int16_t, int16_t, xcb_window_t*, uint32_t*)':
┃        > /build/source/QtScrcpy/util/mousetap/xmousetap.cpp:34:57: error: comparison of integer expressions of different signedness: 'int' and 'uint32_t' {aka 'unsigned int'} [-Werror=sign-compare]
┃        >    34 |             if (!*grab_window || gg->width * gg->height <= *grab_window_size) {
┃        >       |                                  ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
┃        > At global scope:
┃        > cc1plus: note: unrecognized command-line option '-Wno-nested-anon-types' may have been intended to silence earlier diagnostics
┃        > cc1plus: all warnings being treated as errors
┃        > make[2]: *** [QtScrcpy/CMakeFiles/QtScrcpy.dir/build.make:248: QtScrcpy/CMakeFiles/QtScrcpy.dir/util/mousetap/xmousetap.cpp.o] Error 1
┃        > make[2]: *** Waiting for unfinished jobs....
┃        > make[1]: *** [CMakeFiles/Makefile2:121: QtScrcpy/CMakeFiles/QtScrcpy.dir/all] Error 2
┃        > make: *** [Makefile:91: all] Error 2
┃        For full logs, run 'nix log /nix/store/q5h1dib9s3pqx1473fl02a9yvfp0av9w-qt-scrcpy-2.2.1.drv'.
```

**Comment by Daru-san** [comment]
I found a way to get around it using 
```nix
{  
  env.NIX_CFLAGS_COMPILE = toString [
    "-Wno-error=sign-compare"
  ];
}
```

**Comment by Daru-san** [comment]
But now I am stumped by this:
```
make: *** No rule to make target 'install'.  Stop.
```

**Comment by Daru-san** [comment]
Build successful :smile:

File: pkgs/by-name/qt/qt-scrcpy/package.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  ];
+
+  env.NIX_CFLAGS_COMPILE = toString [
+    "-Wno-error=sign-compare"
+  ];
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin
+    cp -r ../output/x64/Release/* $out/bin
```
Comment (myclevorname) [inline pkgs/by-name/qt/qt-scrcpy/package.nix:40]: Doesn't this automatically happen?

Comment (Aleksanaa) [reply pkgs/by-name/qt/qt-scrcpy/package.nix:40]: I suppose it should be if we are not overriding `installPhase`. Also `runHook postInstall` is missing here.

File: pkgs/by-name/qt/qt-scrcpy/package.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  '';
+
+  meta = with lib; {
```
Comment (Aleksanaa) [inline pkgs/by-name/qt/qt-scrcpy/package.nix:43]:
```suggestion
  meta = {
```

File: pkgs/by-name/qt/qt-scrcpy/package.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    cmake
+  ];
+
+  buildInputs = with libsForQt5; [
+    qtbase
+    wrapQtAppsHook
+    qtmultimedia
+    scrcpy
+    qtx11extras
+  ];
```
Comment (Aleksanaa) [inline pkgs/by-name/qt/qt-scrcpy/package.nix:31]:
```suggestion
  nativeBuildInputs = [
    cmake
    libsForQt5.wrapQtAppsHook
  ];

  buildInputs = [ scrcpy ] ++ (with libsForQt5; [
    qtbase
    qtmultimedia
    qtx11extras
  ]);
```

(Please apply formatting)

**Review by Aleksanaa — COMMENTED** [review]
The attribute name and pname should be `qtscrcpy`

File: pkgs/by-name/qt/qt-scrcpy/package.nix (hunk 1 above)
Comment (Daru-san) [reply pkgs/by-name/qt/qt-scrcpy/package.nix]: I think it should happen automatically, but it seems there is no make install target in the `CmakeLists.txt` file, so the manual route works

**Comment by Daru-san** [comment]
There is an issue preventing the build from happening, it affects my system and github CI. Apparently the submodule `QtScrcpyCore`, from the repo https://github.com/barry-ran/QtScrcpyCore cannot be cloned, giving this error
```
┃        > Cloning into '/nix/store/dkb5z7p60vy3l52lv42i5jiib374skyb-source/QtScrcpy/QtScrcpyCore'...
┃        > error: cannot run ssh: No such file or directory
┃        > fatal: unable to fork
┃        > fatal: clone of 'git@github.com:barry-ran/QtScrcpyCore.git' into submodule path '/nix/store/dkb5z7p60vy3l52lv42i5jiib374skyb-source/QtScrcpy/QtScrcpyCore' fail…
┃        > Failed to clone 'QtScrcpy/QtScrcpyCore'. Retry scheduled
┃        > Cloning into '/nix/store/dkb5z7p60vy3l52lv42i5jiib374skyb-source/QtScrcpy/QtScrcpyCore'...
┃        > error: cannot run ssh: No such file or directory
┃        > fatal: unable to fork
┃        > fatal: clone of 'git@github.com:barry-ran/QtScrcpyCore.git' into submodule path '/nix/store/dkb5z7p60vy3l52lv42i5jiib374skyb-source/QtScrcpy/QtScrcpyCore' fail…
┃        > Failed to clone 'QtScrcpy/QtScrcpyCore' a second time, aborting
┃        For full logs, run 'nix log /nix/store/nfa9swir54ydyv38f3c2yrvfgb1zki8w-source.drv'.
```

**Comment by Aleksanaa** [comment]
> There is an issue preventing the build from happening, it affects my system and github CI. Apparently the submodule `QtScrcpyCore`, from the repo [barry-ran/QtScrcpyCore](https://github.com/barry-ran/QtScrcpyCore) cannot be cloned, giving this error

See https://github.com/NixOS/nixpkgs/issues/195117. The best option here would be manually copying that repo to store, since we don't have a lot of submodules. I'll try to fix `fetchgit`.

**Comment by Daru-san** [comment]
> > There is an issue preventing the build from happening, it affects my system and github CI. Apparently the submodule `QtScrcpyCore`, from the repo [barry-ran/QtScrcpyCore](https://github.com/barry-ran/QtScrcpyCore) cannot be cloned, giving this error
> 
> See #195117. The best option here would be manually copying that repo to store, since we don't have a lot of submodules. I'll try to fix `fetchgit`.
> 
> edit: [#195117 (comment)](https://github.com/NixOS/nixpkgs/issues/195117#issuecomment-1410398050) would be fine.

Cool, the solution in the provided issue works perfectly. My question is just if this would be a temporary fix to keep until the issue has been resolved properly

**Review by myclevorname — APPROVED** [review]

**Comment by Aleksanaa** [comment]
There's vendor blob hidden in https://github.com/barry-ran/QtScrcpyCore/tree/5f0359053fb67f2226fcc87919aee0de17d2079b/src/third_party/ffmpeg/lib, which makes it unable to build on other architectures (and we should avoid unneeded blobs), see ofBorg https://logs.ofborg.org/?key=nixos/nixpkgs.340257&attempt_id=fe9242e4-8d56-41ff-8b7f-b817305c2b7f. There should be some options lying somewhere in cmake files. Do you think you can handle it? I can help you if you can't.

**Comment by Aleksanaa** [comment]
I'm working on it. (Anyway I feel that this behavior is actually a bit malicious. Maybe it is really doing some bad things secretly)

**Comment by Daru-san** [comment]
Oh I definitely did not realize that, I don't really get what it may be doing, I need to do more research into that, _package compilation as a whole_. But I would love to understand what it's doing, if you could explain.

**Comment by Aleksanaa** [comment]
> I don't really get what it may be doing, I need to do more research into that

tldr. it's not only vendoring ffmpeg blobs, but also scrcpy-server and adb in git submodules. This is very suspicious, because adb is a privileged operation for android, and scrcpy-server will be installed on the android system. Some users may not notice these pre-built binaries hidden in this open source project.

**Comment by myclevorname** [comment]
Should we close the PR?

**Comment by Aleksanaa** [comment]
Not necessarily, we can also try to replace the blob

**Comment by Daru-san** [comment]
I know this question may not be helpful, but how would you go about replacing the blobs?

I assume you would somehow patch the source or _maybe_ fork the repo and manually replace them.

**Comment by Aleksanaa** [comment]
Resolved

**Comment by Daru-san** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 340257`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>qtscrcpy</li>
  </ul>
</details>

**Comment by emilazy** [comment]
Does this package only support FFmpeg 4? We are trying to get rid of it and this is the only user other than ROCm and the binary packages for Spotify and REAPER.

**Comment by Aleksanaa** [comment]
The upstream vendored this version, so I replaced it with the same package. I can try to see if it still compiles after replacing it with a higher version
